### PR TITLE
Envoy gloo bump for inja escaped characters

### PR DESCRIPTION
**Pull request by jbohanon**
# Description

This gets the latest envoy-gloo in order to test functionality with new Inja feature and new custom callback

Control plane changes to accommodate this new functionality:
Settings CRD: 
```yaml
glooOptions:
  transformationOptions:
    escapeCharacters: true
```

TransformationStages:
Setting can be configured as part of `stagedTransformations` on [VirtualHostOptions](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options.proto.sk/#virtualhostoptions), [RouteOptions](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options.proto.sk/#routeoptions), or [WeightedDestinationOptions](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options.proto.sk/#weighteddestinationoptions)

# Context

See context [here](https://github.com/solo-io/envoy-gloo/pull/258)

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/main/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: docs/content/guides/traffic_management/request_processing/transformations/_index.md [hunk 1]
```diff
@@ -398,6 +402,8 @@ In addition to the standard functions available in the core _Inja_ library, you
 - `base64_encode(string)`: encodes the input string to base64.
 - `base64_decode(string)`: decodes the input string from base64.
 - `substring(string, start_pos, substring_len)`: returns a substring of the input string, starting at `start_pos` and extending for `substring_len` characters. If no `substring_len` is provided or `substring_len` is <= 0, the substring extends to the end of the input string.
+- `replace_with_random(string, pattern)`: returns the input string with instances matching the `pattern` replaced with random characters.
+- `raw_string(string)`: returns the input string with escaped characters intact. Useful for constructing JSON response bodies.
```
Comment (jbohanon) [inline docs/content/guides/traffic_management/request_processing/transformations/_index.md:406]: nit:
```suggestion
- `raw_string(string)`: returns the input string with escaped characters intact. Useful for constructing JSON request or response bodies.
```

File: docs/content/guides/traffic_management/request_processing/transformations/_index.md [hunk 2]
```diff
@@ -380,6 +381,9 @@ This attribute determines which notation to use when accessing elements in JSON
 
 Please note that, if set to `true`, you will need to use the `extraction` function to access extractors in template strings (e.g. `{{ extraction("myExtractor") }}`); if the default value of `false` is used, extractors will simply be available by their name (e.g. `{{ myExtractor }}`).
 
+##### escapeCharacters
+This attribute is used to set for the entire transformation whether Inja should be configured to preserved escaped characters in strings. This is particularly useful when using context from the request or response body to construct new JSON bodies.
```
Comment (jbohanon) [inline docs/content/guides/traffic_management/request_processing/transformations/_index.md:385]: nit:
```suggestion
This attribute is used to set for the entire transformation whether Inja should be configured to preserve escaped characters in strings. This is particularly useful when using context from the request or response body to construct new JSON bodies.
```

File: projects/gloo/api/external/envoy/extensions/transformation/transformation.proto [hunk 3]
```diff
@@ -283,6 +283,10 @@ message TransformationTemplate {
   }
   // Use this field to set Dynamic Metadata.
   repeated DynamicMetadataValue dynamic_metadata_values = 9;
+
+  // Use this field to set Inja behavior when rendering strings which contain
+  // characters that would need to be escaped to be valid JSON
+  bool escape_characters = 12;
```
Comment (jbohanon) [inline projects/gloo/api/external/envoy/extensions/transformation/transformation.proto:289]:
```suggestion
  // Use this field to set Inja behavior when rendering strings which contain
  // characters that would need to be escaped to be valid JSON. Note that this
  // sets the behavior for the entire transformation. Use raw_strings function
  // for fine-grained control within a template.
  bool escape_characters = 12;
```

File: test/e2e/route_transformation_test.go [hunk 4]
```diff
@@ -71,13 +71,16 @@ var _ = Describe("Transformations", func() {
 			}
 		})
 
+		defaultPostBody := "{\"body\":\"test\"}"
```
Comment (jbohanon) [inline test/e2e/route_transformation_test.go:74]: nit:
```suggestion
		defaultPostBody := `{"body":"test"}`
```

**Review by nfuden — CHANGES_REQUESTED** [review]
We should allow for inheritance either via a gateway or settings level default

File: projects/gloo/api/external/envoy/extensions/transformation/transformation.proto (hunk 3 above)
Comment (nfuden) [reply projects/gloo/api/external/envoy/extensions/transformation/transformation.proto:289]: Can we make this a nullable bool? that way we can do inheritance more easily

File: projects/gloo/api/v1/settings.proto [hunk 5]
```diff
@@ -693,11 +693,20 @@ message GlooOptions {
...
+        // transformations are applied and it is difficult to determine which are
+        // causing issues. Defaults to false.
+        google.protobuf.BoolValue log_transformation_request_response_info = 1;
+        // Set escapeCharacters for all TransformationTemplates on all vhosts and routes.
+        // This setting can be overridden in individual TransformationTemplates
+        google.protobuf.BoolValue escape_characters = 2;
+    }
+
+    // Transformation options for entire Gloo instance
+    TransformationOptions transformation_options = 16;
```
Comment (elcasteel) [inline projects/gloo/api/v1/settings.proto:709]: I think this is a breaking change to change `16` from a bool to a TransformationOptions

Comment (jbohanon) [reply projects/gloo/api/v1/settings.proto:709]: It is, but the field we're changing was introduced in an earlier 1.15 beta

Comment (nfuden) [reply projects/gloo/api/v1/settings.proto:709]: wasnt it backported to 1.14 though?

Comment (jbohanon) [reply projects/gloo/api/v1/settings.proto:709]: Sure was... that's a bummer. I'll get that changed

Comment (jbohanon) [reply projects/gloo/api/v1/settings.proto:709]: Sure was... that's a bummer. I'll get that changed

**Review by elcasteel — APPROVED** [review]
LGTM!

**Review by elcasteel — DISMISSED** [review]

File: projects/gloo/pkg/plugins/transformation/plugin.go [hunk 6]
```diff
@@ -338,6 +345,102 @@ func TranslateTransformation(glooTransform *transformation.Transformation) (
...
+			outTemplate.GetHeaders()[k] = &envoytransformation.InjaTemplate{Text: v.GetText()}
+		}
+	}
+
+	if len(inTemplate.GetHeadersToAppend()) > 0 {
+		outTemplate.HeadersToAppend = make(
+			[]*envoytransformation.TransformationTemplate_HeaderToAppend,
+			len(inTemplate.GetHeadersToAppend()))
+		headers := inTemplate.GetHeadersToAppend()
+		for i := range headers {
```
Comment (nfuden) [inline projects/gloo/pkg/plugins/transformation/plugin.go:400]: for i, hdr := range headers is more idiomatic and only an extra allocation

Comment (jbohanon) [reply projects/gloo/pkg/plugins/transformation/plugin.go:400]: It's one allocation, but it copies  the slice val out to the temp on every loop iteration

**Review by nfuden — APPROVED** [review]

**Review by nfuden — APPROVED** [review]

**Review by elcasteel — APPROVED** [review]

**Pull request merged**